File: cmps/12B/labs/lab5/Dictionary.c
// Ilay Raz
// ilraz
// CMPS12M-02
// 2/27/18
// Dictionary.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "Dictionary.h"


// Private types -----------------------------

typedef struct NodeObj {
    char* key;
    char* value;
    struct NodeObj* next;
} NodeObj;

typedef NodeObj* Node;

Node newNode(char* key, char* value) {
    Node node = malloc(sizeof *node);
    assert(node != NULL);
    node->key = key;
    node->value = value;
    node->next = NULL;
    return node;
}

void freeNode(Node* node) {
    if (node != NULL && *node != NULL) {
        free(*node);
        *node = NULL;
    }
}

typedef struct DictionaryObj {
    Node head;
    int size;
} DictionaryObj;

// public functions -----------------------------

Dictionary newDictionary(void) {
    Dictionary dict = malloc(sizeof *dict);
    assert(dict != NULL);
    dict->head = NULL;
    dict->size = 0;
    return dict;
}

void freeDictionary(Dictionary* dict) {
    if (dict != NULL && *dict != NULL) {
        makeEmpty(*dict);
        free(*dict);
        *dict = NULL;
    }
}

int isEmpty(Dictionary dict) {
    return dict->size == 0;
}

int size(Dictionary dict) {
    return dict->size;
}

char* lookup(Dictionary dict, char* key) {
    for (Node node = dict->head; node != NULL; node = node->next)
        if (strcmp(node->key, key) == 0)
            return node->value;
    return NULL;
}

void insert(Dictionary dict, char* key, char* value) {
    assert(lookup(dict, key) == NULL);
    if (dict->head == NULL)
        dict->head = newNode(key, value);
    else {
        Node node = dict->head;
        for (; node->next != NULL; node = node->next);
        node->next = newNode(key, value);
    }
    dict->size++;
}

void delete(Dictionary dict, char* key) {
    assert(dict->head != NULL);
    Node pNode, dNode;
    pNode = dict->head;

    if (strcmp(pNode->key, key) == 0) {
        dict->head = dict->head->next;
        freeNode(&pNode);
    } else {
        for (; pNode->next != NULL && strcmp(pNode->next->key, key) != 0; pNode = pNode->next);
        assert(pNode->next != NULL);
        dNode = pNode->next;
        if (dNode == NULL)
            pNode->next = NULL;
        else
            pNode->next = dNode->next;
        freeNode(&dNode);
    }
    dict->size--;
}

void makeEmpty(Dictionary dict) {
    Node temp;
    while (dict->head != NULL) {
        temp = dict->head;
        dict->head = dict->head->next;
        freeNode(&temp);
    }
    dict->size = 0;
}

void printDictionary(FILE* out, Dictionary dict) {
    for (Node node = dict->head; node != NULL; node = node->next)
        fprintf(out, "%s %s\n", node->key, node->value);
}

File: cmps/12B/assignments/ass3/Dictionary.java
// Ilay Raz
// ilraz
// CMPS12B-02
// 2/6/18
// Dictionary.java

public class Dictionary implements DictionaryInterface {

    private class Node {
        public String key;
        public String value;
        public Node next;
        public Node parent;

        public Node(String myKey, String myValue) {
            key = myKey;
            value = myValue;
            next = null;
            parent = null;
        }

        public Node(String myKey, String myValue, Node myParent) {
            key = myKey;
            value = myValue;
            next = null;
            parent = myParent;
        }
    }

    private int size;
    private Node head;
    private Node tail;

    public Dictionary() {
        size = 0;
        head = null;
        tail = head;
    }

    private Node findKey(String key) {
        for (Node node = head; node != null; node = node.next)
            if (node.key == key)
                return node;
        return null;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public int size() {
        return size;
    }

    public String lookup(String key) {
        Node node = findKey(key);
        return node == null ? null : node.value;
    }

    public void insert(String key, String value) throws DuplicateKeyException {
        if (lookup(key) != null)
            throw new DuplicateKeyException("cannot insert duplicate keys");
        if (size == 0) {
            head = new Node(key, value);
            tail = head;
            size++;
        } else {
            tail.next = new Node(key, value, tail);
            tail = tail.next;
            size++;
        }
    }

    public void delete(String key) throws KeyNotFoundException {
        Node node = findKey(key);
        if (node == null)
            throw new KeyNotFoundException("cannot delete non existent key");
        size--;
        if (size == 0)
            head = tail = null;
        else if (node == head)
            head = head.next;
        else if (node == tail) {
            tail = tail.parent;
            tail.next = null;
        } else {
            node.parent.next = node.next;
            node.next.parent = node.parent;
        }
    }

    public void makeEmpty() {
        head = tail = null;
        size = 0;
    }

    public String toString() {
        String string = "";
        for (Node node = head; node != null; node = node.next)
            string += node.key + " " + node.value + "\n";
        return string;
    }
}
